Common/src/main/java/com/blamejared/crafttweaker/api/tag/manager/ITagManager.java

```diff
@@ -1,187 +1,158 @@
-public interface ITagManager<T> extends CommandStringDisplayable {
-    
-    @Nonnull
-    Class<T> getElementClass();
-    
-    /**
-     * Retrieves a tag by its name.
-     * Will also be called by the BEP.
-     * <p>
-     * Note that this method does _not_ yet create the tag if it does not exist.
-     * Adding something to the object created by this tag will create it for the game.
-     *
-     * @param name The Resource location of the tag
-     *
-     * @return A Tag object.
-     */
+@Document("vanilla/api/tag/manager/ITagManager")
+@ZenCodeType.Name("crafttweaker.api.tag.manager.ITagManager")
+public interface ITagManager<T> extends CommandStringDisplayable, Iterable<MCTag<T>> {
+
+    @ZenCodeType.Method
+    @ZenCodeType.Getter("tagFolder")
+    default String tagFolder() {
+
+        String tagDir = TagManager.getTagDir(resourceKey());
+
+        // Really not ideal, but I don't see a better way, lets just hope that other mods don't be dumb and add their tags to other folders.
+        if (tagDir.startsWith("tags/")) {
+            tagDir = tagDir.substring("tags/".length());
+        }
+        return tagDir;
+    }
+
     @ZenCodeType.Method
-    default MCTag<T> getTag(String name) {
-        
-        return getTag(new ResourceLocation(name));
+    default MCTag<T> tag(String id) {
+
+        ResourceLocation rl = ResourceLocation.tryParse(id);
+        Objects.requireNonNull(rl, "Invalid Resource Location passed! Given: '" + id + "'");
+        return tag(rl);
     }
-    
-    /**
-     * Retrieves a tag by its name.
-     * Will also be called by the BEP.
-     * <p>
-     * Note that this method does _not_ yet create the tag if it does not exist.
-     * Adding something to the object created by this tag will create it for the game.
-     *
-     * @param location The Resource location of the tag
-     *
-     * @return A Tag object.
-     */
+
     @ZenCodeType.Method
-    default MCTag<T> getTag(ResourceLocation location) {
-        
-        return new MCTag<>(location, this);
+    default MCTag<T> tag(ResourceLocation id) {
+
+        return tagMap().getOrDefault(id, new MCTag<>(id, this));
     }
-    
-    /**
-     * Checks if a tag already exists. Does the same as calling `.exists` on a tag directly
-     *
-     * @param name The resource location to check for
-     *
-     * @return Whether or not this tag already exists
-     */
+
+
+    @ZenCodeType.Method
+    default void addElements(MCTag<T> to, T... values) {
+
+        if (!exists(to)) {
+            CraftTweakerAPI.apply(new ActionTagCreate<>(to, List.of(values)));
+        } else {
+            CraftTweakerAPI.apply(new ActionTagAdd<>(to, List.of(values)));
+        }
+        recalculate();
+    }
+
+    @ZenCodeType.Method
+    default void removeElements(MCTag<T> from, T... values) {
+
+        if (!exists(from)) {
+            throw new IllegalArgumentException("Cannot remove elements from empty tag: " + from);
```

Comment:
Why an IAE instead of a do-nothing action?

Reply:
Helps with errors, if someone had:
```
<tag:items:minecraft:woool>.remove(<item:minecraft:white_wool>);
```
They won't get an error, saying the tag doesn't exist.